### ISSUE TITLE
how to merge the parallel ops?

### ISSUE THREAD

**Issue by junranhe** [dmlc/tvm#449]
in dense net and inception, there are many small layer from the same input, I want to merge the parallel ops, for load the input once
my code for example like this, I want to merge fused_op for use memory effect, how can I do this by tvm:

def bn_act_conv(from_layer, n_filter):
    bn = batchnorm(from_layer)
    relu = relu(bn)
    return conv_1x1(relu, n_filter)

def fused_op(input):
  conv_0 = bn_act_conv(input, 32) 
  conv_1 = bn_act_conv(input, 32)
  conv_2 = bn_act_conv(input, 32)
  return concat([conv_0, conv_1, conv_2], dim=1)

**Comment by tqchen** [comment]
What you described is essentially a grouped version of conv and then bn. We might be be able to directly declare a group conv and then fuse it with bn and act, then at high level graphir, do the merge of parallel path-> group conv.

**Comment by junranhe** [comment]
thanks for reply,  I think there are some thing different, batch norm layer has channel scale params and bias params, so when the input go throw bn_relu, It will be change to difference input.
so I think bn_act_conv != conv_bn_act :)

**Comment by junranhe** [comment]
@tqchen  I'm sorry to trouble you,have two points to add: 
1) If the num_filter of the small conv are not the same:
conv_0 = bn_act_conv(input, 16)
conv_1 = bn_act_conv(input, 32)
conv_2 = bn_act_conv(input, 64)
It can't declare as a group conv simply.
2) If my ops'type are not the same and just share the input?
def fused_op(inpt):
  conv_0 = bn_act_conv(input, 16)
  pool_1 = max_pool(input)
  other_2 = other_layer(intput)
  return concat([conv_0, pool_1, other_2],dim=1)

**Comment by tqchen** [comment]
I see what you mean, if the group number is regular(e.g. same) I think it still makes sense to merge them into a grouped version of op(although you also have group wise scale shift).

If group numbers or other configurations are not the same, maybe it makes sense to optimize each kernel separately and launch them via multiple stream. Note that while most libraries optimizes for maximum perf of a single kernel. We could use TVM to get different kernels that might be slower when running a single op(utilizing fewer resources), but gives faster throughput when running those kernels concurrently

**Comment by junranhe** [comment]
thanks for reply, I see, I can implem grouped version of op derectly, instead of auto merge the parallel ops by tvm

**Comment by tqchen** [comment]
great, contribution to TOPI is more than welcomed!
